File: sdk/python/src/protomcp/runner.py
import json
import os
import re
import sys
import time

import protomcp.protomcp_pb2 as pb

import inspect

from protomcp.transport import Transport
from protomcp.tool import get_registered_tools, get_hidden_tool_names
from protomcp.result import ToolResult
from protomcp.context import ToolContext, _deliver_sampling_response
from protomcp.log import ServerLogger
from protomcp.middleware import get_registered_middleware
from protomcp import manager
from protomcp.resource import get_registered_resources, get_registered_resource_templates, ResourceContent
from protomcp.prompt import get_registered_prompts
from protomcp.completion import get_completion_handler, CompletionResult
from protomcp.server_context import resolve_contexts
from protomcp.local_middleware import build_middleware_chain
from protomcp.telemetry import emit_telemetry, ToolCallEvent
from protomcp.sidecar import start_sidecars
from protomcp.discovery import discover_handlers, get_config

def _uri_matches_template(template: str, uri: str) -> bool:
    """Check if a URI matches a URI template pattern like 'notes://{id}'."""
    pattern = re.sub(r'\{[^}]+\}', '[^/]+', template)
    return bool(re.fullmatch(pattern, uri))

log: ServerLogger = ServerLogger(send_fn=lambda msg: None)

_first_tool_call = True

def run():
    socket_path = os.environ.get("PROTOMCP_SOCKET")
    if not socket_path:
        print("PROTOMCP_SOCKET not set — run via 'pmcp dev'", file=sys.stderr)
        sys.exit(1)

    transport = Transport(socket_path)
    transport.connect()
    manager._init(transport)

    start_sidecars("server_start")
    discover_handlers()

    global log
    log = ServerLogger(send_fn=transport.send)

    # Track middleware handlers for intercept dispatch
    _mw_handlers = {}

    while True:
        try:
            env = transport.recv()
        except ConnectionError:
            break

        if env.HasField("list_tools"):
            _handle_list_tools(transport, env)
            _send_middleware_registrations(transport, _mw_handlers)
            _disable_hidden_tools(transport)
        elif env.HasField("call_tool"):
            _handle_call_tool(transport, env)
        elif env.HasField("reload"):
            _handle_reload(transport, env, _mw_handlers)
        elif env.HasField("middleware_intercept"):
            _handle_middleware_intercept(transport, env, _mw_handlers)
        elif env.HasField("list_resources_request"):
            _handle_list_resources(transport, env)
        elif env.HasField("read_resource_request"):
            _handle_read_resource(transport, env)
        elif env.HasField("list_resource_templates_request"):
            _handle_list_resource_templates(transport, env)
        elif env.HasField("list_prompts_request"):
            _handle_list_prompts(transport, env)
        elif env.HasField("get_prompt_request"):
            _handle_get_prompt(transport, env)
        elif env.HasField("completion_request"):
            _handle_completion(transport, env)
        elif env.HasField("sampling_response"):
            _deliver_sampling_response(env.request_id, env.sampling_response)

def _handle_list_tools(transport, env):
    tools = get_registered_tools()
    tool_defs = []
    for t in tools:
        tool_defs.append(pb.ToolDefinition(
            name=t.name,
            description=t.description,
            input_schema_json=t.input_schema_json,
            output_schema_json=t.output_schema_json,
            title=t.title,
            destructive_hint=t.destructive_hint,
            idempotent_hint=t.idempotent_hint,
            read_only_hint=t.read_only_hint,
            open_world_hint=t.open_world_hint,
            task_support=t.task_support,
        ))
    resp = pb.Envelope(
        tool_list=pb.ToolListResponse(tools=tool_defs),
        request_id=env.request_id,
    )
    transport.send(resp)

def _handle_call_tool(transport, env):
    global _first_tool_call
    if _first_tool_call:
        _first_tool_call = False
        start_sidecars("first_tool_call")

    req = env.call_tool
    tools = get_registered_tools()
    handler = None
    for t in tools:
        if t.name == req.name:
            handler = t.handler
            break

    if handler is None:
        resp = pb.Envelope(
            call_result=pb.CallToolResponse(
                is_error=True,
                result_json=json.dumps([{"type": "text", "text": f"Tool not found: {req.name}"}]),
            ),
            request_id=env.request_id,
        )
        transport.send(resp)
        return

    start_time = time.monotonic()
    action_name = ""
    try:
        args = json.loads(req.arguments_json) if req.arguments_json else {}
        # Resolve server contexts and inject into args if handler accepts them
        ctx_values = resolve_contexts(args)
        sig = inspect.signature(handler)
        for param_name, value in ctx_values.items():
            if param_name in sig.parameters:
                args[param_name] = value
        # Build middleware chain around handler
        chain = build_middleware_chain(req.name, handler)
        action_name = args.get("action", "")
        emit_telemetry(ToolCallEvent(tool_name=req.name, action=action_name, phase="start", args=dict(args)))
        start_time = time.monotonic()
        if "ctx" in sig.parameters:
            ctx = ToolContext(
                progress_token=req.progress_token,
                send_fn=transport.send,
            )
            result = chain(ctx, args)
        else:
            result = chain(None, args)

        elapsed_ms = int((time.monotonic() - start_time) * 1000)
        result_str = result.result if isinstance(result, ToolResult) else str(result)
        emit_telemetry(ToolCallEvent(tool_name=req.name, action=action_name, phase="success", args={}, result=str(result_str)[:20000], duration_ms=elapsed_ms))

        if isinstance(result, ToolResult):
            resp_msg = pb.CallToolResponse(
                is_error=result.is_error,
                result_json=json.dumps([{"type": "text", "text": str(result.result)}]),
                enable_tools=result.enable_tools or [],
                disable_tools=result.disable_tools or [],
            )
            if result.is_error and result.error_code:
                resp_msg.error.CopyFrom(pb.ToolError(
                    error_code=result.error_code,
                    message=result.message or "",
                    suggestion=result.suggestion or "",
                    retryable=result.retryable,
                ))
        else:
            resp_msg = pb.CallToolResponse(
                result_json=json.dumps([{"type": "text", "text": str(result)}]),
            )
    except Exception as e:
        elapsed_ms = int((time.monotonic() - start_time) * 1000)
        emit_telemetry(ToolCallEvent(tool_name=req.name, action=action_name, phase="error", args={}, error=e, duration_ms=elapsed_ms))
        resp_msg = pb.CallToolResponse(
            is_error=True,
            result_json=json.dumps([{"type": "text", "text": str(e)}]),
        )

    # Check if result_json exceeds chunk threshold — stream if so.
    chunk_threshold = int(os.environ.get('PROTOMCP_CHUNK_THRESHOLD', '65536'))
    result_json_str = resp_msg.result_json
    result_json_bytes = result_json_str.encode('utf-8') if result_json_str else b''

    if len(result_json_bytes) > chunk_threshold:
        transport.send_raw(
            request_id=env.request_id,
            field_name='result_json',
            data=result_json_bytes,
        )
    else:
        resp = pb.Envelope(call_result=resp_msg, request_id=env.request_id)
        transport.send(resp)

def _handle_reload(transport, env, mw_handlers):
    # Re-discover handlers if hot_reload is enabled
    if get_config().get("hot_reload"):
        discover_handlers()
    # For now, just acknowledge. Full reload with importlib is complex.
    resp = pb.Envelope(
        reload_response=pb.ReloadResponse(success=True),
        request_id=env.request_id,
    )
    transport.send(resp)
    # Also re-send tool list with empty request_id so it routes to handshakeCh
    fake_env = pb.Envelope()  # empty envelope (request_id defaults to "")
    _handle_list_tools(transport, fake_env)
    _send_middleware_registrations(transport, mw_handlers)

def _send_middleware_registrations(transport, mw_handlers):
    mw_defs = get_registered_middleware()
    for mw in mw_defs:
        mw_handlers[mw.name] = mw.handler
        reg = pb.Envelope(
            register_middleware=pb.RegisterMiddlewareRequest(
                name=mw.name,
                priority=mw.priority,
            ),
        )
        transport.send(reg)
        # Wait for acknowledgment
        try:
            ack = transport.recv()
        except ConnectionError:
            return
    # Send handshake-complete signal
    complete = pb.Envelope(
        reload_response=pb.ReloadResponse(success=True),
    )
    transport.send(complete)

def _disable_hidden_tools(transport):
    """Disable any tools registered with hidden=True after handshake."""
    hidden = get_hidden_tool_names()
    if not hidden:
        return
    env = pb.Envelope(disable_tools=pb.DisableToolsRequest(tool_names=hidden))
    transport.send(env)

def _handle_middleware_intercept(transport, env, mw_handlers):
    req = env.middleware_intercept
    handler = mw_handlers.get(req.middleware_name)

    resp_fields = {}
    if handler:
        try:
            result = handler(
                phase=req.phase,
                tool_name=req.tool_name,
                args_json=req.arguments_json,
                result_json=req.result_json,
                is_error=req.is_error,
            )
            if isinstance(result, dict):
                resp_fields = result
        except Exception as e:
            resp_fields = {"reject": True, "reject_reason": str(e)}

    resp = pb.Envelope(
        middleware_intercept_response=pb.MiddlewareInterceptResponse(
            arguments_json=resp_fields.get("arguments_json", ""),
            result_json=resp_fields.get("result_json", ""),
            reject=resp_fields.get("reject", False),
            reject_reason=resp_fields.get("reject_reason", ""),
        ),
        request_id=env.request_id,
    )
    transport.send(resp)

def _handle_list_resources(transport, env):
    resources = get_registered_resources()
    defs = [pb.ResourceDefinition(
        uri=r.uri, name=r.name, description=r.description,
        mime_type=r.mime_type, size=r.size,
    ) for r in resources]
    resp = pb.Envelope(
        resource_list_response=pb.ResourceListResponse(resources=defs),
        request_id=env.request_id,
    )
    transport.send(resp)

def _handle_list_resource_templates(transport, env):
    templates = get_registered_resource_templates()
    defs = [pb.ResourceTemplateDefinition(
        uri_template=t.uri_template, name=t.name,
        description=t.description, mime_type=t.mime_type,
    ) for t in templates]
    resp = pb.Envelope(
        resource_template_list_response=pb.ResourceTemplateListResponse(templates=defs),
        request_id=env.request_id,
    )
    transport.send(resp)

def _handle_read_resource(transport, env):
    uri = env.read_resource_request.uri
    resources = get_registered_resources()
    templates = get_registered_resource_templates()

    handler = None
    for r in resources:
        if r.uri == uri:
            handler = r.handler
            break
    if handler is None:
        for t in templates:
            if _uri_matches_template(t.uri_template, uri):
                handler = t.handler
                break

    if handler is None:
        resp = pb.Envelope(
            read_resource_response=pb.ReadResourceResponse(contents=[
                pb.ResourceContent(uri=uri, text=f"Resource not found: {uri}", mime_type="text/plain")
            ]),
            request_id=env.request_id,
        )
        transport.send(resp)
        return

    try:
        result = handler(uri)
        if isinstance(result, ResourceContent):
            result = [result]
        elif isinstance(result, str):
            result = [ResourceContent(uri=uri, text=result)]

        contents = []
        for rc in result:
            c = pb.ResourceContent(uri=rc.uri, mime_type=rc.mime_type)
            if rc.blob:
                c.blob = rc.blob
            else:
                c.text = rc.text
            contents.append(c)

        resp = pb.Envelope(
            read_resource_response=pb.ReadResourceResponse(contents=contents),
            request_id=env.request_id,
        )
    except Exception as e:
        resp = pb.Envelope(
            read_resource_response=pb.ReadResourceResponse(contents=[
                pb.ResourceContent(uri=uri, text=str(e), mime_type="text/plain")
            ]),
            request_id=env.request_id,
        )
    transport.send(resp)

def _handle_list_prompts(transport, env):
    prompts = get_registered_prompts()
    defs = []
    for p in prompts:
        args = [pb.PromptArgument(name=a.name, description=a.description, required=a.required) for a in p.arguments]
        defs.append(pb.PromptDefinition(name=p.name, description=p.description, arguments=args))
    resp = pb.Envelope(
        prompt_list_response=pb.PromptListResponse(prompts=defs),
        request_id=env.request_id,
    )
    transport.send(resp)

def _handle_get_prompt(transport, env):
    req = env.get_prompt_request
    prompts = get_registered_prompts()
    handler = None
    for p in prompts:
        if p.name == req.name:
            handler = p.handler
            break

    if handler is None:
        resp = pb.Envelope(
            get_prompt_response=pb.GetPromptResponse(
                description=f"Prompt not found: {req.name}",
                messages=[pb.PromptMessage(role="assistant", content_json=json.dumps({"type": "text", "text": f"Prompt not found: {req.name}"}))],
            ),
            request_id=env.request_id,
        )
        transport.send(resp)
        return

    try:
        args = json.loads(req.arguments_json) if req.arguments_json else {}
        result = handler(**args)
        if not isinstance(result, list):
            result = [result]

        messages = []
        for msg in result:
            content_json = json.dumps({"type": "text", "text": msg.content})
            messages.append(pb.PromptMessage(role=msg.role, content_json=content_json))

        resp = pb.Envelope(
            get_prompt_response=pb.GetPromptResponse(description="", messages=messages),
            request_id=env.request_id,
        )
    except Exception as e:
        resp = pb.Envelope(
            get_prompt_response=pb.GetPromptResponse(
                description="Error",
                messages=[pb.PromptMessage(role="assistant", content_json=json.dumps({"type": "text", "text": str(e)}))],
            ),
            request_id=env.request_id,
        )
    transport.send(resp)

def _handle_completion(transport, env):
    req = env.completion_request
    handler = get_completion_handler(req.ref_type, req.ref_name, req.argument_name)

    if handler is None:
        resp = pb.Envelope(
            completion_response=pb.CompletionResponse(values=[], total=0, has_more=False),
            request_id=env.request_id,
        )
        transport.send(resp)
        return

    try:
        result = handler(req.argument_value)
        if isinstance(result, CompletionResult):
            resp = pb.Envelope(
                completion_response=pb.CompletionResponse(
                    values=result.values, total=result.total, has_more=result.has_more,
                ),
                request_id=env.request_id,
            )
        elif isinstance(result, list):
            resp = pb.Envelope(
                completion_response=pb.CompletionResponse(values=result, total=len(result), has_more=False),
                request_id=env.request_id,
            )
        else:
            resp = pb.Envelope(
                completion_response=pb.CompletionResponse(values=[], total=0, has_more=False),
                request_id=env.request_id,
            )
    except Exception as e:
        resp = pb.Envelope(
            completion_response=pb.CompletionResponse(values=[], total=0, has_more=False),
            request_id=env.request_id,
        )
    transport.send(resp)

File: sdk/go/protomcp/runner.go
package protomcp

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"

	pb "github.com/msilverblatt/protomcp/gen/proto/protomcp"
)

var firstToolCallOnce sync.Once

var Log *ServerLogger

// mwHandlers maps middleware name to handler during runtime.
var mwHandlers map[string]func(phase, toolName, argsJSON, resultJSON string, isError bool) map[string]interface{}

// activeCancels maps request_id to cancel function for in-flight tool calls.
var activeCancels map[string]context.CancelFunc
var cancelsMu sync.Mutex

func Run() {
	socketPath := os.Getenv("PROTOMCP_SOCKET")
	if socketPath == "" {
		fmt.Fprintln(os.Stderr, "PROTOMCP_SOCKET not set — run via 'pmcp dev'")
		os.Exit(1)
	}

	tp := NewTransport(socketPath)
	if err := tp.Connect(); err != nil {
		fmt.Fprintf(os.Stderr, "connect: %v\n", err)
		os.Exit(1)
	}
	defer tp.Close()

	Log = NewServerLogger(func(env *pb.Envelope) error { return tp.Send(env) }, "protomcp-go")
	mwHandlers = make(map[string]func(phase, toolName, argsJSON, resultJSON string, isError bool) map[string]interface{})
	activeCancels = make(map[string]context.CancelFunc)

	StartSidecars("server_start")
	defer StopAllSidecars()

	for {
		env, err := tp.Recv()
		if err != nil {
			break
		}

		reqID := env.GetRequestId()

		switch {
		case env.GetListTools() != nil:
			handleListTools(tp, reqID)
			sendMiddlewareRegistrations(tp)
			sendDisableHiddenTools(tp)
		case env.GetCallTool() != nil:
			handleCallTool(tp, env.GetCallTool(), reqID)
		case env.GetReload() != nil:
			handleReload(tp, reqID)
		case env.GetMiddlewareIntercept() != nil:
			handleMiddlewareIntercept(tp, env.GetMiddlewareIntercept(), reqID)
		case env.GetListResourcesRequest() != nil:
			handleListResources(tp, reqID)
		case env.GetListResourceTemplatesRequest() != nil:
			handleListResourceTemplates(tp, reqID)
		case env.GetReadResourceRequest() != nil:
			handleReadResource(tp, env.GetReadResourceRequest(), reqID)
		case env.GetListPromptsRequest() != nil:
			handleListPrompts(tp, reqID)
		case env.GetGetPromptRequest() != nil:
			handleGetPrompt(tp, env.GetGetPromptRequest(), reqID)
		case env.GetCompletionRequest() != nil:
			handleCompletion(tp, env.GetCompletionRequest(), reqID)
		case env.GetCancel() != nil:
			cancelsMu.Lock()
			if cancel, ok := activeCancels[env.GetCancel().GetRequestId()]; ok {
				cancel()
			}
			cancelsMu.Unlock()
		}
	}
}

func handleListTools(tp *Transport, reqID string) {
	tools := GetRegisteredTools()
	var defs []*pb.ToolDefinition
	for _, t := range tools {
		defs = append(defs, &pb.ToolDefinition{
			Name:            t.Name,
			Description:     t.Desc,
			InputSchemaJson: t.InputSchemaJSON(),
			DestructiveHint: t.Destructive,
			IdempotentHint:  t.Idempotent,
			ReadOnlyHint:    t.ReadOnly,
			OpenWorldHint:   t.OpenWorld,
			TaskSupport:     t.TaskSupport,
		})
	}
	tp.Send(&pb.Envelope{
		RequestId: reqID,
		Msg: &pb.Envelope_ToolList{
			ToolList: &pb.ToolListResponse{Tools: defs},
		},
	})
}

func sendDisableHiddenTools(tp *Transport) {
	var hidden []string
	for _, t := range GetRegisteredTools() {
		if t.Hidden {
			hidden = append(hidden, t.Name)
		}
	}
	if len(hidden) == 0 {
		return
	}
	tp.Send(&pb.Envelope{
		Msg: &pb.Envelope_DisableTools{
			DisableTools: &pb.DisableToolsRequest{ToolNames: hidden},
		},
	})
}

func sendHandshakeComplete(tp *Transport) {
	tp.Send(&pb.Envelope{
		Msg: &pb.Envelope_ReloadResponse{
			ReloadResponse: &pb.ReloadResponse{Success: true},
		},
	})
}

func sendMiddlewareRegistrations(tp *Transport) {
	mws := GetRegisteredMiddleware()
	for _, mw := range mws {
		mwHandlers[mw.Name] = mw.Handler
		tp.Send(&pb.Envelope{
			Msg: &pb.Envelope_RegisterMiddleware{
				RegisterMiddleware: &pb.RegisterMiddlewareRequest{
					Name:     mw.Name,
					Priority: mw.Priority,
				},
			},
		})
		// Wait for acknowledgment
		if _, err := tp.Recv(); err != nil {
			return
		}
	}
	sendHandshakeComplete(tp)
}

func buildResultJSON(text string) string {
	content := []map[string]string{{"type": "text", "text": text}}
	data, err := json.Marshal(content)
	if err != nil {
		return `[{"type":"text","text":""}]`
	}
	return string(data)
}

func handleCallTool(tp *Transport, req *pb.CallToolRequest, reqID string) {
	tools := GetRegisteredTools()
	var handler func(ToolContext, map[string]interface{}) ToolResult
	for _, t := range tools {
		if t.Name == req.Name {
			handler = t.HandlerFn
			break
		}
	}

	if handler == nil {
		tp.Send(&pb.Envelope{
			RequestId: reqID,
			Msg: &pb.Envelope_CallResult{
				CallResult: &pb.CallToolResponse{
					IsError:    true,
					ResultJson: buildResultJSON("Tool not found: " + req.Name),
				},
			},
		})
		return
	}

	var args map[string]interface{}
	if req.ArgumentsJson != "" {
		if err := json.Unmarshal([]byte(req.ArgumentsJson), &args); err != nil {
			tp.Send(&pb.Envelope{
				RequestId: reqID,
				Msg: &pb.Envelope_CallResult{
					CallResult: &pb.CallToolResponse{
						IsError:    true,
						ResultJson: buildResultJSON("Invalid arguments JSON: " + err.Error()),
						Error: &pb.ToolError{
							ErrorCode:  "INVALID_INPUT",
							Message:    fmt.Sprintf("Failed to parse arguments: %s", err.Error()),
							Suggestion: "Ensure arguments is valid JSON",
							Retryable:  false,
						},
					},
				},
			})
			return
		}
	}
	if args == nil {
		args = map[string]interface{}{}
	}

	// Start first_tool_call sidecars on first invocation
	firstToolCallOnce.Do(func() { StartSidecars("first_tool_call") })

	// Resolve server contexts
	ctxValues := ResolveContexts(args)
	for k, v := range ctxValues {
		args[k] = v
	}

	callCtx, cancel := context.WithCancel(context.Background())
	cancelsMu.Lock()
	activeCancels[reqID] = cancel
	cancelsMu.Unlock()

	ctx := ToolContext{
		Ctx:           callCtx,
		ProgressToken: req.ProgressToken,
		sendFn:        func(env *pb.Envelope) error { return tp.Send(env) },
	}

	// Emit telemetry start
	EmitTelemetry(ToolCallEvent{
		ToolName: req.Name,
		Phase:    "start",
		Args:     args,
	})

	startTime := time.Now()

	// Build local middleware chain
	wrapped := BuildMiddlewareChain(req.Name, handler)
	result := wrapped(ctx, args)

	durationMs := int(time.Since(startTime).Milliseconds())

	// Emit telemetry success/error
	if result.IsError {
		EmitTelemetry(ToolCallEvent{
			ToolName:   req.Name,
			Phase:      "error",
			Args:       args,
			Result:     result.ResultText,
			DurationMs: durationMs,
			Message:    result.Message,
		})
	} else {
		EmitTelemetry(ToolCallEvent{
			ToolName:   req.Name,
			Phase:      "success",
			Args:       args,
			Result:     result.ResultText,
			DurationMs: durationMs,
		})
	}

	cancel()
	cancelsMu.Lock()
	delete(activeCancels, reqID)
	cancelsMu.Unlock()

	resp := &pb.CallToolResponse{
		IsError:      result.IsError,
		ResultJson:   buildResultJSON(result.ResultText),
		EnableTools:  result.EnableTools,
		DisableTools: result.DisableTools,
	}
	if result.IsError && result.ErrorCode != "" {
		resp.Error = &pb.ToolError{
			ErrorCode:  result.ErrorCode,
			Message:    result.Message,
			Suggestion: result.Suggestion,
			Retryable:  result.Retryable,
		}
	}

	// Use raw sideband for large payloads to avoid protobuf overhead.
	chunkThreshold := 65536
	if v := os.Getenv("PROTOMCP_CHUNK_THRESHOLD"); v != "" {
		if n, err := strconv.Atoi(v); err == nil {
			chunkThreshold = n
		}
	}

	resultBytes := []byte(resp.ResultJson)
	if len(resultBytes) > chunkThreshold {
		tp.SendRaw(reqID, "result_json", resultBytes)
	} else {
		tp.Send(&pb.Envelope{
			RequestId: reqID,
			Msg:       &pb.Envelope_CallResult{CallResult: resp},
		})
	}
}

func handleReload(tp *Transport, reqID string) {
	tp.Send(&pb.Envelope{
		RequestId: reqID,
		Msg: &pb.Envelope_ReloadResponse{
			ReloadResponse: &pb.ReloadResponse{Success: true},
		},
	})
	handleListTools(tp, "")
	sendMiddlewareRegistrations(tp)
}

func uriMatchesTemplate(template, uri string) bool {
	parts := strings.Split(template, "{")
	if len(parts) == 0 {
		return template == uri
	}
	if !strings.HasPrefix(uri, parts[0]) {
		return false
	}
	remaining := uri[len(parts[0]):]
	for _, part := range parts[1:] {
		closeBrace := strings.Index(part, "}")
		if closeBrace < 0 {
			return false
		}
		suffix := part[closeBrace+1:]
		if suffix == "" {
			if remaining == "" {
				return false
			}
			remaining = ""
		} else {
			idx := strings.Index(remaining, suffix)
			if idx <= 0 {
				return false
			}
			remaining = remaining[idx+len(suffix):]
		}
	}
	return remaining == ""
}

func handleListResources(tp *Transport, reqID string) {
	resources := GetRegisteredResources()
	var defs []*pb.ResourceDefinition
	for _, r := range resources {
		defs = append(defs, &pb.ResourceDefinition{
			Uri:         r.URI,
			Name:        r.Name,
			Description: r.Description,
			MimeType:    r.MimeType,
			Size:        r.Size,
		})
	}
	tp.Send(&pb.Envelope{
		RequestId: reqID,
		Msg: &pb.Envelope_ResourceListResponse{
			ResourceListResponse: &pb.ResourceListResponse{Resources: defs},
		},
	})
}

func handleListResourceTemplates(tp *Transport, reqID string) {
	templates := GetRegisteredResourceTemplates()
	var defs []*pb.ResourceTemplateDefinition
	for _, t := range templates {
		defs = append(defs, &pb.ResourceTemplateDefinition{
			UriTemplate: t.URITemplate,
			Name:        t.Name,
			Description: t.Description,
			MimeType:    t.MimeType,
		})
	}
	tp.Send(&pb.Envelope{
		RequestId: reqID,
		Msg: &pb.Envelope_ResourceTemplateListResponse{
			ResourceTemplateListResponse: &pb.ResourceTemplateListResponse{Templates: defs},
		},
	})
}

func handleReadResource(tp *Transport, req *pb.ReadResourceRequest, reqID string) {
	uri := req.GetUri()

	// Try static resources first.
	for _, r := range GetRegisteredResources() {
		if r.URI == uri && r.HandlerFn != nil {
			contents := r.HandlerFn()
			sendResourceContents(tp, reqID, contents)
			return
		}
	}

	// Try resource templates.
	for _, t := range GetRegisteredResourceTemplates() {
		if t.HandlerFn != nil && uriMatchesTemplate(t.URITemplate, uri) {
			contents := t.HandlerFn(uri)
			if len(contents) > 0 {
				sendResourceContents(tp, reqID, contents)
				return
			}
		}
	}

	// No matching resource found.
	tp.Send(&pb.Envelope{
		RequestId: reqID,
		Msg: &pb.Envelope_ReadResourceResponse{
			ReadResourceResponse: &pb.ReadResourceResponse{},
		},
	})
}

func sendResourceContents(tp *Transport, reqID string, contents []ResourceContent) {
	var pbContents []*pb.ResourceContent
	for _, c := range contents {
		pbContents = append(pbContents, &pb.ResourceContent{
			Uri:      c.URI,
			MimeType: c.MimeType,
			Text:     c.Text,
			Blob:     c.Blob,
		})
	}
	tp.Send(&pb.Envelope{
		RequestId: reqID,
		Msg: &pb.Envelope_ReadResourceResponse{
			ReadResourceResponse: &pb.ReadResourceResponse{Contents: pbContents},
		},
	})
}

func handleListPrompts(tp *Transport, reqID string) {
	prompts := GetRegisteredPrompts()
	var defs []*pb.PromptDefinition
	for _, p := range prompts {
		var args []*pb.PromptArgument
		for _, a := range p.Arguments {
			args = append(args, &pb.PromptArgument{
				Name:        a.Name,
				Description: a.Description,
				Required:    a.Required,
			})
		}
		defs = append(defs, &pb.PromptDefinition{
			Name:        p.Name,
			Description: p.Description,
			Arguments:   args,
		})
	}
	tp.Send(&pb.Envelope{
		RequestId: reqID,
		Msg: &pb.Envelope_PromptListResponse{
			PromptListResponse: &pb.PromptListResponse{Prompts: defs},
		},
	})
}

func handleGetPrompt(tp *Transport, req *pb.GetPromptRequest, reqID string) {
	prompts := GetRegisteredPrompts()
	for _, p := range prompts {
		if p.Name == req.GetName() && p.HandlerFn != nil {
			var args map[string]string
			if req.GetArgumentsJson() != "" {
				if err := json.Unmarshal([]byte(req.GetArgumentsJson()), &args); err != nil {
					args = map[string]string{}
				}
			}
			if args == nil {
				args = map[string]string{}
			}
			desc, messages := p.HandlerFn(args)
			var pbMsgs []*pb.PromptMessage
			for _, m := range messages {
				pbMsgs = append(pbMsgs, &pb.PromptMessage{
					Role:        m.Role,
					ContentJson: m.ContentJSON,
				})
			}
			tp.Send(&pb.Envelope{
				RequestId: reqID,
				Msg: &pb.Envelope_GetPromptResponse{
					GetPromptResponse: &pb.GetPromptResponse{
						Description: desc,
						Messages:    pbMsgs,
					},
				},
			})
			return
		}
	}
	// Prompt not found — return empty response.
	tp.Send(&pb.Envelope{
		RequestId: reqID,
		Msg: &pb.Envelope_GetPromptResponse{
			GetPromptResponse: &pb.GetPromptResponse{},
		},
	})
}

func handleCompletion(tp *Transport, req *pb.CompletionRequest, reqID string) {
	handler, ok := GetCompletionHandler(req.GetRefType(), req.GetRefName(), req.GetArgumentName())
	if !ok || handler == nil {
		tp.Send(&pb.Envelope{
			RequestId: reqID,
			Msg: &pb.Envelope_CompletionResponse{
				CompletionResponse: &pb.CompletionResponse{},
			},
		})
		return
	}
	result := handler(req.GetArgumentValue())
	tp.Send(&pb.Envelope{
		RequestId: reqID,
		Msg: &pb.Envelope_CompletionResponse{
			CompletionResponse: &pb.CompletionResponse{
				Values:  result.Values,
				Total:   result.Total,
				HasMore: result.HasMore,
			},
		},
	})
}

func handleMiddlewareIntercept(tp *Transport, req *pb.MiddlewareInterceptRequest, reqID string) {
	resp := &pb.MiddlewareInterceptResponse{
		ArgumentsJson: req.ArgumentsJson,
		ResultJson:    req.ResultJson,
	}

	handler, ok := mwHandlers[req.MiddlewareName]
	if ok && handler != nil {
		result := handler(req.Phase, req.ToolName, req.ArgumentsJson, req.ResultJson, req.IsError)
		if result != nil {
			if v, ok := result["reject"].(bool); ok && v {
				resp.Reject = true
				if r, ok := result["reject_reason"].(string); ok {
					resp.RejectReason = r
				}
			}
			if v, ok := result["arguments_json"].(string); ok {
				resp.ArgumentsJson = v
			}
			if v, ok := result["result_json"].(string); ok {
				resp.ResultJson = v
			}
		}
	}

	tp.Send(&pb.Envelope{
		RequestId: reqID,
		Msg:       &pb.Envelope_MiddlewareInterceptResponse{MiddlewareInterceptResponse: resp},
	})
}
